src/Renderer.php

```diff
@@ -206,6 +209,97 @@ public function inlineJson($values)
+     *
+     * @param  string $content the html in the textblock stripe as a string
+     * @param  int    $index   the index of the stripe within the page
+     * @return array
+     *
+     */
+    private function getBreakStripeHeading($stripe, $index) {
+        if (empty($stripe['heading']['html'])) {
+            return null;
+        } else {
+            return [
+                'level' => 1,
+                'headingText' => strip_tags($stripe['heading']['html']),
+                'id' => 'vssl-stripe--break--heading-' . $index
+            ];
+        }
+    }
+
+    /**
+     * Given a textblock stripe's content and index within the page, returns an array of list items for rendering in a table of contents
+     *
+     * @param  string $content the html in the textblock stripe as a string
+     * @param  int    $index   the index of the stripe within the page
+     * @return array
+     *
+     */
+    private function getTextblockStripeHeadings($stripe, $index) {
+        if (empty($stripe['content']['html'])) {
+            return null;
+        }
+
+        $dom = new DOMDocument();
+        $dom->loadHTML($stripe['content']['html']);
+        $xpath = new DOMXPath($dom);
+        $headers = $xpath->query('//h1 | //h2');
+
+        $listItems = [];
+        $h1Count = 0;
+        $h2Count = 0;
+        foreach ($headers as $header) {
+            $tagName = $header->tagName;
+            $prefix = 'vssl-stripe--textblock--heading-';
+            $headerIndex = $tagName === 'h1' ? $h1Count : $h2Count;
+            $id = $prefix . $index . '-' . $tagName . '-' . $headerIndex;
+
+            array_push($listItems, [
+                'level' => $tagName === 'h1' ? 1 : 2,
+                'headingText' => $header->nodeValue,
+                'id' => $id
+            ]);
+
+            if ($tagName === 'h1') $h1Count++;
+            else $h2Count++;
+        }
+        return $listItems;
+    }
+
+    /**
+     * Returns a table of contents based on the contents of the page
+     *
+     * @return string
+     */
+    public function getTableOfContents($scope)
+    {
+        if (empty($this->data['stripes'])) {
+            return null;
+        }
+
+        $stripes = $this->data['stripes'];
+        $items = array_map(function ($stripe, $index) use ($scope) {
+            if (
+                (empty($scope) || strpos($scope, 'break') !== false) &&
+                $stripe['type'] === 'stripe-break'
+            ) {
+                return [$this->getBreakStripeHeading($stripe, $index)];
+            }
+            if (
+                (empty($scope) || strpos($scope, 'textblock') !== false) &&
+                $stripe['type'] === 'stripe-textblock'
+            ) {
+                return $this->getTextblockStripeHeadings($stripe, $index);
+            }
+        }, $stripes, array_keys($stripes));
+
+        $flatItems = array_merge(...array_filter($items));
```

Comment:
Ah—remind me why you're doing this like this, rather than just `$flatItems = array_filter($items)`?

Reply:
here `$items` is an array where each value is either null or an array of arrays.
 
so `array_filter($items)` to remove the `nulls` and then `array_merge(...args)` to merge the nested arrays together.

so that `$flatItems` is an array of arrays. 

(I think I'm using array correctly here, but the bottom child array in each of these is the actual fields of data for the heading and the parent arrays are indexes with values. Not sure if there is a distinction in PHP)

Reply:
All that said, maybe pushing to an array in a foreach loop makes more sense here...

Reply:
yeah this is much better

```php
        $stripes = $this->data['stripes'];
        $items = [];
        foreach ($stripes as $index => $stripe) {
            if (
                (empty($scope) || strpos($scope, 'break') !== false) &&
                $stripe['type'] === 'stripe-break'
            ) {
                array_push($items, $this->getBreakStripeHeading($stripe, $index));
            }
            if (
                (empty($scope) || strpos($scope, 'textblock') !== false) &&
                $stripe['type'] === 'stripe-textblock'
            ) {
                array_push($items, ...$this->getTextblockStripeHeadings($stripe, $index));
            }
        };

        return $items;
```